modules/tinymce/src/core/main/ts/fmt/FormatChanged.ts

```diff
@@ -5,101 +5,172 @@
-  const parents = Arr.filter(FormatUtils.getParents(editor.dom, elm), (node) => node.nodeType === 1 && !node.getAttribute('data-mce-bogus'));
-
-  // Check for new formats
-  Obj.each(formatChangeData, (data: FormatData, format: string) => {
-    Tools.each(parents, (node: Node) => {
-      if (editor.formatter.matchNode(node, format, {}, data.similar)) {
-        if (formatsList.indexOf(format) === -1) {
-          // Execute callbacks
-          Arr.each(data.callbacks, (callback: FormatChangeCallback) => {
-            callback(true, { node, format, parents });
-          });
-
-          newFormats[format] = data.callbacks;
-        }
+  editor.on('FormatApply FormatRemove', (e: EditorEvent<FormatEvent>) => {
+    const element = Optional.from(e.node)
+      .map((nodeOrRange) => Obj.get(nodeOrRange as RangeLikeObject, 'startContainer').getOr(nodeOrRange as Node))
+      .bind((node) => NodeType.isElement(node) ? Optional.some(node) : Optional.from(node.parentElement))
+      .getOrThunk(() => editor.selection.getNode());
 
-        matchedFormats[format] = data.callbacks;
-        return false;
-      }
+    updateAndFireChangeCallbacks(editor, element, registeredFormatListeners.get());
+  });
+};
 
-      if (MatchFormat.matchesUnInheritedFormatSelector(editor, node, format)) {
+const matchingNode = (editor: Editor, parents: Element[], format: string, similar: boolean, vars?: FormatVars): Optional<Element> => {
+  const isMatchingNode = (node: Element) => {
+    const matchingFormat = editor.formatter.matchNode(node, format, vars ?? {}, similar);
+    return !Type.isUndefined(matchingFormat);
+  };
+  const isUnableToMatch = (node: Element) => {
+    if (MatchFormat.matchesUnInheritedFormatSelector(editor, node, format)) {
+      return true;
+    } else {
+      if (!similar) {
+        // If we want to find an exact match, then finding a similar match halfway up the parents tree is bad
+        return Type.isNonNullable(editor.formatter.matchNode(node, format, vars, true));
+      } else {
         return false;
       }
-    });
-  });
-
-  // Check if current formats still match
-  const remainingFormats = filterRemainingFormats(currentFormats.get(), matchedFormats, elm, parents);
-
-  // Update the current formats
-  currentFormats.set({
-    ...newFormats,
-    ...remainingFormats
-  });
+    }
+  };
+  return Arr.findUntil(parents, isMatchingNode, isUnableToMatch);
 };
 
-const filterRemainingFormats = (currentFormats: FormatCallbacks, matchedFormats: FormatCallbacks, elm: Element, parents: Node[]) => Obj.bifilter(currentFormats, (callbacks: FormatChangeCallback[], format: string) => {
-  if (!Obj.has(matchedFormats, format)) {
-    // Execute callbacks
-    Arr.each(callbacks, (callback: FormatChangeCallback) => {
-      callback(false, { node: elm, format, parents });
-    });
+const getParents = (editor: Editor, elm?: Element): Element[] => {
+  const element = elm ?? editor.selection.getNode();
```

Comment:
Similar to the above, I believe using `getNode` is incorrect here and we want to be using `getStart` instead since the old logic also always worked off that instead of the common ancestor.

Reply:
Also maybe this fallback should be a shared function between the `FormatApply` lookup and here?

Reply:
Changed to use getStart, but I don't think it's worth a shared function for just the one line of code

Reply:
That's not why I wanted this as a shared function, it's that this behaviour should always be the same. So if someone comes along later and changes one they are quite likely to miss the second case.

Reply:
Done